docs/contributing/writing-instrumentation-module.md

```diff
@@ -360,3 +360,138 @@ For example:
+a variable that is local to the advice execution for communication between the enter and exit advices.
+
+When advices are not inlined, usage of `@Advice.Local` is not possible. It is however possible to
+return a value from the enter advice and get the value in the exit advice with a parameter annotated
+with `@Advice.Enter`, for example:
+
+```java
+@Advice.OnMethodEnter(suppress = Throwable.class)
+public static Object onEnter(@Advice.Argument(1) Object request) {
+  return "enterValue";
+}
+
+@Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class)
+public static void onExit(@Advice.Argument(1) Object request,
+                          @Advice.Enter Object enterValue) {
+  // do something with enterValue
+}
+```
+
+### modifying method arguments
+
+With inlined advices, using the `@Advice.Argument` annotation on method parameter with `readOnly = false`
+allows to modify instrumented method arguments.
+
+When using non-inlined advices, reading the argument values is still done with `@Advice.Argument`
+annotated parameters, however modifying the values is done through the advice method return value
+and `@Advice.AssignReturned.ToArguments` annotation:
+
+```java
+@Advice.OnMethodEnter(suppress = Throwable.class)
+@Advice.AssignReturned.ToArguments(@ToArgument(1))
+public static Object onEnter(@Advice.Argument(1) Object request) {
+  return "hello";
+}
+```
+
+It is possible to modify multiple arguments at once by using an array, see usages of
+`@Advice.AssignReturned.ToArguments` for detailed examples.
+
+### modifying method return value
+
+With inlined advices, using the `@Advice.Return` annotation on method parameter with `readOnly = false`
+allows to modify instrumented method return value on exit advice.
+
+When using non-inlined advices, reading the original return value is still done with `@Advice.Return`
+annotated parameter, however modifying the value is done through the advice method return value
+and `@Advice.AssignReturned.ToReturned`
+
+```java
+@Advice.OnMethodExit(suppress = Throwable.class)
+@Advice.AssignReturned.ToReturned
+public static Object onEnter(@Advice.Return Object returnValue) {
+  return "hello";
+}
+```
+
+### writing to internal class fields
+
+With inlined advices, using the `@Advice.FieldValue(value = "fieldName", readOnly = false)` annotation
+on advice method parameter allows to modify the `fieldName` field of the instrumented class.
+
+When using non-inlined advices, reading the original field value is still done with `@Advice.FieldValue`
+annotated parameter, however modifying the value is done through the advice method return value
+and `@Advice.AssignReturned.ToFields` annotation:
+
+```java
+@Advice.OnMethodExit(suppress = Throwable.class)
+@Advice.AssignReturned.ToFields(@ToField(value = "fieldName"))
+public static Object onEnter(@Advice.FieldValue("fieldName") Object originalFieldValue) {
```

Comment:
```suggestion
public static Object onExit(@Advice.FieldValue("fieldName") Object originalFieldValue) {
```

Reply:
field modification usually happens before the original method body is executed, is there any particular reason to switch to method exit ?

Reply:
Ah, gotcha. I suggested this change to make it match the annotation used for this function (`OnMethodExit`). Though based on what you mentioned, maybe the annotation should be changed to `OnMethodEnter`?

Reply:
Good catch, I completely missed that (obvious) one !!